data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimap;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimaps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeSet;
+import org.apache.iceberg.util.StructProjection;
+import org.apache.parquet.Preconditions;
+
+public abstract class DeleteFilter<T> {
+  private static final long DEFAULT_SET_FILTER_THRESHOLD = 100_000L;
+  private static final Schema POS_DELETE_SCHEMA = new Schema(
+      MetadataColumns.DELETE_FILE_PATH,
+      MetadataColumns.DELETE_FILE_POS);
+
+  private final long setFilterThreshold;
+  private final FileIO io;
+  private final DataFile dataFile;
+  private final List<DeleteFile> posDeletes;
+  private final List<DeleteFile> eqDeletes;
+  private final Schema requiredSchema;
+  private final Accessor<StructLike> posAccessor;
+
+  public DeleteFilter(FileIO io, FileScanTask task, Schema tableSchema, Schema requestedSchema) {
+    this.setFilterThreshold = DEFAULT_SET_FILTER_THRESHOLD;
+    this.io = io;
+    this.dataFile = task.file();
+
+    ImmutableList.Builder<DeleteFile> posDeleteBuilder = ImmutableList.builder();
+    ImmutableList.Builder<DeleteFile> eqDeleteBuilder = ImmutableList.builder();
+    for (DeleteFile delete : task.deletes()) {
+      switch (delete.content()) {
+        case POSITION_DELETES:
+          posDeleteBuilder.add(delete);
+          break;
+        case EQUALITY_DELETES:
+          eqDeleteBuilder.add(delete);
+          break;
+        default:
+          throw new UnsupportedOperationException("Unknown delete file content: " + delete.content());
+      }
+    }
+
+    this.posDeletes = posDeleteBuilder.build();
+    this.eqDeletes = eqDeleteBuilder.build();
+    this.requiredSchema = fileProjection(tableSchema, requestedSchema, posDeletes, eqDeletes);
+    this.posAccessor = requiredSchema.accessorForField(MetadataColumns.ROW_POSITION.fieldId());
+  }
+
+  public Schema requiredSchema() {
+    return requiredSchema;
+  }
+
+  Accessor<StructLike> posAccessor() {
+    return posAccessor;
+  }
+
+  protected abstract StructLike asStructLike(T record);
+
+  protected long pos(T record) {
+    return (Long) posAccessor.get(asStructLike(record));
+  }
+
+  public CloseableIterable<T> filter(CloseableIterable<T> records) {
+    return applyEqDeletes(applyPosDeletes(records));
```

Comment:
Is there anything important about the order of applying deletes here? Is the guess here that there will be more Pos Deletes than EqDeletes?

Reply:
My thought here is just that EqDelete always uses a set check so it's probably going to be cheaper than the possibility that you have to do the streaming check in pos deletes, but maybe I'm thinking about it wrong.

Reply:
We might want to do minor compaction to transform the equality delete files to position delete files, so I guess position deletes should be more than equality deletes as time goes on.

The records in position deletes are ordered, so the sorted merge-base check wouldn't be more expensive than set based check consider the cost of building the hash set. Right?

Reply:
I think that equality deletes are more expensive to apply because they require a projection and a set lookup (hash and maybe equality check), and there could be multiple equality deletes to apply. So the idea is to do the cheapest operation first and the most expensive operation last to do fewer expensive filter checks.

Reply:
> do minor compaction to transform the equality delete files to position delete files

@chenjunjiedada  do we really need to transform equality-deletes to positional-deletes when doing minor compaction ?    Let's take the case in a single bucket: 

txn-0:  insert-file0,  pos-delete-file0,  equality-delete-file0;
txn-1:  insert-file1,  pos-delete-file1,   equality-delete-file1;
txn-2:  insert-file2,  pos-delete-file2,  equality-delete-file2;

The insert-file0's posDeletes is `[pos-delete-file0, pos-delete-file1, pos-delete-file2]` and eqDeletes is `[equality-delete-file0, equality-delete-file1, equality-delete-file2]`. 

The insert-file1's posDeletes is `[pos-delete-file1, pos-delete-file2]` and eqDeletes is `[equality-delete-file1, equality-delete-file2]`

The insert-file2's posDeletes is `[pos-delete-file2]` and eqDeletes is `[equality-delete-file2]` 

You mean we will transform the equality-delete-file2 and equality-delete-file1 to a new pos-delete-file3 when doing minior compaciton for txn2 ?

Reply:
@openinx , Since the cost of merging equality delete file is bigger than merging position delete, so it could be an option for minor compaction. While I haven't think about how we will do that compaction.  I guess we should consider the sequence number when compacting the equality delete file, for example, convert equality delete files with same sequence number to a new position delete file.